### PR TITLE
Nullable and Modernization for Stride.Core.Serialization

### PR DESCRIPTION
**Pull request by IXLLEGACYIXL**
# PR Details

Modernizing Stride.Core.Reflection 

## Description

1. Should enable Nullable instead of Strides NotNull attribute
2. Should reduce the usage of reflection, mainly used in collection Descriptors ( thats now prototyped for Dictionary, if its in interested i can convert set and list and the others too )
3. Using new C# features ie collection initialization, Pattern Matching etc

## Related Issue

#2156 
 #2155
Removes all [NotNull] annotations and uses the new Nullable 
Removes a bug taht System.Collections anyway crash the editor as serializers are missing

## Types of changes

- [x] Docs change / refactoring / dependency upgrade ( some comments changed  )
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] **I have built and run the editor to try this change out.**

### CODE REVIEW

File: sources/core/Stride.Core.Reflection/MemberPath.cs [hunk 1]
```diff
@@ -379,46 +379,6 @@ public bool TryGetValue(object rootObject, out object value)
             return true;
         }
 
-        ///// <summary>
```
Comment (IXLLEGACYIXL) [inline sources/core/Stride.Core.Reflection/MemberPath.cs:382]: thats just zombie code, no idea what its for

File: sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs [hunk 2]
```diff
@@ -39,17 +39,17 @@ public PropertyDescriptor(ITypeDescriptor typeDescriptor, PropertyInfo propertyI
...
         }
 
-        public override void Set(object thisObject, object value)
+        public override void Set(object thisObject, object? value)
         {
             if (!HasSet)
                 throw new InvalidOperationException($"The property [{Name}] of type [{DeclaringType.Name}] has no setter.");
 
-            setMethod.Invoke(thisObject, new[] {value});
+            setMethod!.Invoke(thisObject, [value]);
```
Comment (Eideren) [inline sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs:52]: I think this would be cleaner as 
```cs
if (setMethod is null)
    throw new InvalidOperationException($"The property [{Name}] of type [{DeclaringType.Name}] has no setter.");

setMethod.Invoke(thisObject, [value]);
```
That way, if the code ever changes, which is very unlikely but still, nullability would still be validated for the invoke instead of always ignored

File: sources/core/Stride.Core.Reflection/MemberPath.cs [hunk 3]
```diff
@@ -357,17 +357,17 @@ public object GetValue(object rootObject)
...
+            ArgumentNullException.ThrowIfNull(rootObject);
             value = null;
             try
             {
                 object nextObject = rootObject;
                 for (int i = 0; i < items.Count; i++)
                 {
                     var item = items[i];
-                    nextObject = item.GetValue(nextObject);
+                    nextObject = item.GetValue(nextObject)!;
```
Comment (Eideren) [inline sources/core/Stride.Core.Reflection/MemberPath.cs:370]: Wouldn't it be valid for this method to return both true *and* output null `value` at the same time ?
In which case, this `!` should be removed I think ?

File: sources/core/Stride.Core.Reflection/TypeDescriptors/ObjectDescriptor.cs [hunk 4]
```diff
@@ -144,7 +142,7 @@ public virtual void Initialize(IComparer<object> keyComparer)
             }
 
             // Free the member list
-            members = memberList.ToArray();
+            members = [.. memberList];
```
Comment (Eideren) [inline sources/core/Stride.Core.Reflection/TypeDescriptors/ObjectDescriptor.cs:145]: The new syntax is cool, but maybe this is a bit too much, let's keep `ToArray()`, it is more readable

File: sources/core/Stride.Core.Reflection/TypeDescriptors/PrimitiveDescriptor.cs [hunk 5]
```diff
@@ -12,14 +12,15 @@ namespace Stride.Core.Reflection
...
-        private readonly Dictionary<string, object> enumRemap;
+        private static readonly List<IMemberDescriptor> EmptyMembers = [];
+        private readonly Dictionary<string, object?> enumRemap;
 
         public PrimitiveDescriptor(ITypeDescriptorFactory factory, Type type, bool emitDefaultValues, IMemberNamingConvention namingConvention)
             : base(factory, type, emitDefaultValues, namingConvention)
         {
             if (!IsPrimitive(type))
                 throw new ArgumentException("Type [{0}] is not a primitive");
+            enumRemap = new Dictionary<string, object?>(StringComparer.OrdinalIgnoreCase);
```
Comment (Eideren) [inline sources/core/Stride.Core.Reflection/TypeDescriptors/PrimitiveDescriptor.cs:23]: Looking at the rest of the logic, it looks like you can rollback this change and mark `enumRemap` as nullable ?

File: sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs [hunk 6]
```diff
@@ -11,40 +12,50 @@ namespace Stride.Core.Reflection
...
-        private readonly MethodInfo isReadOnlyMethod;
-
+        private static readonly List<string> ListOfMembersToRemove = ["Comparer", "Capacity"];
+
+        private Action<object, object?> AddMethod;
+        private Action<object, object?> RemoveMethod;
+        private Action<object> ClearMethod;
+        private Func<object, object?,bool> ContainsMethod;
+        private Func<object, int> CountMethod;
+        private Func<object, bool> IsReadOnlyMethod;
```
Comment (Eideren) [inline sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs:22]: Naming convention, private members start with a lowercase letter, same issue with the other descriptors

File: sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs [hunk 7]
```diff
@@ -199,33 +169,17 @@ public override object GetValue(object set, int index)
...
-                object[] valueParam = new[] { value };
-                if ((bool)containsMethod.Invoke(set, indexParam))
-                {
-                    removeMethod.Invoke(set, indexParam);
-                }
-                if (!(bool)containsMethod.Invoke(set, valueParam))
-                {
-                    addMethod.Invoke(set, valueParam);
-                }
+                Add(set, index);
```
Comment (Eideren) [inline sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs:182]: Is using `{ set, index }` instead of `{ set, value }` a bug on the previous version or a typo in your changes ? Or am I not understanding this function ?

File: sources/core/Stride.Core.Yaml.Tests/DescriptorTests.cs [hunk 8]
```diff
@@ -232,7 +224,7 @@ public void TestDictionaryDescriptor()
         {
             var attributeRegistry = new AttributeRegistry();
             var factory = new TypeDescriptorFactory(attributeRegistry);
-            var descriptor = new DictionaryDescriptor(factory, typeof(Dictionary<int, string>), false,
+            DictionaryDescriptor descriptor = new DictionaryDescriptor(factory, typeof(Dictionary<int, string>), false,
```
Comment (Eideren) [inline sources/core/Stride.Core.Yaml.Tests/DescriptorTests.cs:227]: Unnecessary change, not a big deal, but if you have time

**Review by Eideren — COMMENTED** [review]
Looks great, thanks a bunch for your work @IXLLEGACYIXL

File: sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs (hunk 6 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs:22]: i did that as old collection descriptor had also delegates with pascal case, but ye i can change them to camel

File: sources/core/Stride.Core.Reflection/TypeDescriptorFactory.cs [hunk 9]
```diff
@@ -106,7 +109,7 @@ protected virtual ITypeDescriptor Create(Type type)
             }
             else if (type.IsArray)
             {
-                if (type.GetArrayRank() == 1 && !type.GetElementType().IsArray)
+                if (type.GetArrayRank() == 1 && !type.GetElementType()!.IsArray)
```
Comment (Jklawreszuk) [inline sources/core/Stride.Core.Reflection/TypeDescriptorFactory.cs:112]: Looks great but I noticed you did used null-forgiving operator for Nullable type that might be null. Is it intentional @IXLLEGACYIXL ?

**Review by Jklawreszuk — CHANGES_REQUESTED** [review]

File: sources/core/Stride.Core.Reflection/TypeDescriptorFactory.cs (hunk 9 above)
Comment (Eideren) [reply sources/core/Stride.Core.Reflection/TypeDescriptorFactory.cs:112]: This shouldn't return null when the type is an array, which it is given the if test above :)

Comment (Jklawreszuk) [reply sources/core/Stride.Core.Reflection/TypeDescriptorFactory.cs:112]: Right, makes sense! 😅

**Review by Jklawreszuk — APPROVED** [review]

File: sources/core/Stride.Core.Reflection/TypeDescriptors/PrimitiveDescriptor.cs (hunk 5 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/TypeDescriptors/PrimitiveDescriptor.cs:23]: i thought,

if something is found enumRemap is set in any case
if nothing is found its an empty dictionary which is also valid for all the usages in the class

so later on the null checks can be removed which i forgot, 

its like returning an empty list instead of returning null to be valid in any case

File: sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs (hunk 7 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs:182]: if (simpleSet != null)
is always true as IsSet is true when System.Collection.Generic.ISet is provided
there is no System.Collection.ISet and all the descriptors work like "on instantiation check if type fits and after that trust me"

so the else in the original code is never invoked ever

File: sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs [hunk 10]
```diff
@@ -11,40 +12,50 @@ namespace Stride.Core.Reflection
...
-            ElementType = interfaceType.GetGenericArguments()[0] ?? typeof(object);
-
-            Type[] ArgTypes = { ElementType };
-            addMethod = interfaceType.GetMethod("Add", ArgTypes);
-            removeMethod = type.GetMethod("Remove", ArgTypes);
-            clearMethod = interfaceType.GetMethod("Clear");
-            containsMethod = type.GetMethod("Contains", ArgTypes);
-            countMethod = type.GetProperty("Count").GetGetMethod();
-            isReadOnlyMethod = type.GetInterface(typeof(ICollection<>)).GetProperty("IsReadOnly").GetGetMethod();
+                throw new ArgumentException(@"Expecting a type inheriting from System.Collections.ISet<T>", nameof(type));
```
Comment (IXLLEGACYIXL) [inline sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs:31]: this was just copy paste from the other descriptors like dictionary and list, it seems to be added after the others and implemented the same way even though it only works for generic one as the non generic one doesnt exist or isnt handled

File: sources/core/Stride.Core.Reflection/MemberPath.cs (hunk 3 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/MemberPath.cs]: is removed

File: sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs (hunk 2 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs]: now its nullable

File: sources/core/Stride.Core.Reflection/TypeDescriptors/ObjectDescriptor.cs (hunk 4 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/TypeDescriptors/ObjectDescriptor.cs]: reverted

File: sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs (hunk 6 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs]: changed and simplified names to be constant accross the classes

File: sources/core/Stride.Core.Yaml.Tests/DescriptorTests.cs (hunk 8 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Yaml.Tests/DescriptorTests.cs]: reverted that

File: sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs (hunk 2 above)
Comment (Eideren) [reply sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs]: Well, now you just made it more confusing ... I would rather you use your previous one or my suggestion, sorry :(

File: sources/core/Stride.Core.Reflection/MemberPath.cs [hunk 11]
```diff
@@ -357,17 +357,20 @@ public bool Apply(object rootObject, MemberPathAction actionType, object value)
...
                 {
                     var item = items[i];
-                    nextObject = item.GetValue(nextObject)!;
+                    if(nextObject is not null)
+                    {
+                        nextObject = item.GetValue(nextObject);
+                    }
                 }
                 value = nextObject;
             }
```
Comment (Eideren) [inline sources/core/Stride.Core.Reflection/MemberPath.cs]: How about:
```cs
try
{
    object nextObject = rootObject;
    for (int i = 0; i < items.Count; i++)
    {
        var item = items[i];
        var newObject = item.GetValue(nextObject);
        if (newObject is null)
        {
        	value = null;
            return true;
        }
        nextObject = newObject;
    }
    value = nextObject;
}
```

File: sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs (hunk 2 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs]: oh now i get it

File: sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs (hunk 7 above)
Comment (Eideren) [reply sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs:182]: Yep, I get that part, what I was referring to was this section from the previous logic:
```cs
if (simpleSet.Contains(index))
    simpleSet.Remove(index);
if (!simpleSet.Contains(value))
    simpleSet.Add(value);
```
Which, from my understanding, directly maps to your:
```cs
if (Contains(set, index))
    Remove(set, index);
if(!Contains(set, index))
    Add(set, index);
```

Notice the difference between the versions of the second `if` test, the previous version passes the `value` parameter (`!simpleSet.Contains(value)`) while your version uses the `index` parameter `!Contains(set, index)`
Is that intended ?

Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/TypeDescriptors/SetDescriptor.cs]: [8fc1f8b](https://github.com/stride3d/stride/pull/2266/commits/8fc1f8bb5d3534b47e9043106213a6839d0d240f)
made it back to what it was before

File: sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs (hunk 2 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/MemberDescriptors/PropertyDescriptor.cs]: [854cbfb](https://github.com/stride3d/stride/pull/2266/commits/854cbfb8f7a74fba9d9ae797c3c523946f9ca6db)
made it now to your suggestion

File: sources/core/Stride.Core.Reflection/MemberPath.cs (hunk 11 above)
Comment (IXLLEGACYIXL) [reply sources/core/Stride.Core.Reflection/MemberPath.cs]: [291b3ea](https://github.com/stride3d/stride/pull/2266/commits/291b3ea507e619f175b1913538f70aff024ad851)
did that

**Comment by IXLLEGACYIXL** [comment]
ran tests and they fail currently, maybe its a compilation issue... lets see..

**Comment by IXLLEGACYIXL** [comment]
WELL  sample tests fail but on all branches with latest main, but the descriptor tests and all other engien tests pass so it seems good

**Comment by Eideren** [comment]
Yep, it's good, TC reported 4 failing tests but those are unrelated, see #2296

**Comment by Eideren** [comment]
Thanks !
